app/src/main/java/com/duckduckgo/app/browser/BrowserWebViewClient.kt

```diff
@@ -268,12 +269,26 @@ class BrowserWebViewClient @Inject constructor(
         return true
     }
 
-    private fun determineNavigationOptions(webView: WebView): BrowserNavigationOptions {
-        val canGoBack = webView.canGoBack()
-        val canGoForward = webView.canGoForward()
-        return BrowserNavigationOptions(canGoBack, canGoForward)
+    interface BrowserNavigationOptions {
+        val stepsToPreviousPage: Int
+        val canGoBack: Boolean
+        val canGoForward: Boolean
+        val hasNavigationHistory: Boolean
     }
 
-    data class BrowserNavigationOptions(val canGoBack: Boolean, val canGoForward: Boolean)
+    data class WebViewNavigationOptions(val stack: WebBackForwardList): BrowserNavigationOptions {
 
+        override val stepsToPreviousPage: Int = if (stack.isHttpsUpgrade) 2 else 1
+        override val canGoBack: Boolean = stack.currentIndex >= stepsToPreviousPage
+        override val canGoForward: Boolean = stack.currentIndex + 1 < stack.size
+        override val hasNavigationHistory = stack.size != 0
+
+        private val WebBackForwardList.isHttpsUpgrade: Boolean
+            get() {
+                if (currentIndex < 1) return false
+                val current = currentItem.originalUrl
```

Comment:
`currentItem` is flagged as `nullable`

Reply:
Ahh, so it is. I've checked the list size so it definitely exists, I'll make that more clear.